app/src/main/java/pm/gnosis/heimdall/ui/tokens/balances/TokenBalancesFragment.kt

```diff
@@ -0,0 +1,111 @@
+package pm.gnosis.heimdall.ui.tokens.balances
+
+import android.content.Intent
+import android.os.Bundle
+import android.support.v7.widget.LinearLayoutManager
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import com.jakewharton.rxbinding2.support.v4.widget.refreshes
+import com.jakewharton.rxbinding2.view.clicks
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.rxkotlin.plusAssign
+import io.reactivex.rxkotlin.subscribeBy
+import kotlinx.android.synthetic.main.layout_token_balances.*
+import pm.gnosis.heimdall.R
+import pm.gnosis.heimdall.common.di.components.ApplicationComponent
+import pm.gnosis.heimdall.common.di.components.DaggerViewComponent
+import pm.gnosis.heimdall.common.di.modules.ViewModule
+import pm.gnosis.heimdall.common.utils.*
+import pm.gnosis.heimdall.ui.base.Adapter
+import pm.gnosis.heimdall.ui.base.BaseFragment
+import pm.gnosis.heimdall.ui.tokens.add.AddTokenActivity
+import pm.gnosis.heimdall.utils.errorSnackbar
+import pm.gnosis.heimdall.utils.handleQrCodeActivityResult
+import pm.gnosis.utils.hexAsEthereumAddressOrNull
+import pm.gnosis.utils.isValidEthereumAddress
+import timber.log.Timber
+import javax.inject.Inject
+
+class TokenBalancesFragment : BaseFragment() {
+    @Inject lateinit var viewModel: TokenBalancesContract
+    @Inject lateinit var adapter: TokenBalancesAdapter
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
+            inflater.inflate(R.layout.layout_token_balances, container, false)
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        val addressArgument = arguments?.getString(ARGUMENT_ADDRESS)?.hexAsEthereumAddressOrNull()
+        layout_tokens_fab.visibility = if (addressArgument == null) View.VISIBLE else View.GONE
+
+        if (addressArgument != null) {
+            viewModel.setup(addressArgument)
+        } else {
+            snackbar(layout_tokens_coordinator_layout, R.string.invalid_ethereum_address)
```

Comment:
We should finish the activity here, else the app might crash, since the view model will access the address wbich is not initialized